packages/core/src/utils/CollectLabels.ts

```diff
@@ -90,34 +116,44 @@ export const retrieveLabel = (
 // https://stackoverflow.com/a/44564236
 export const collectLabels = async (
   allShapes: Shape[]
-): Promise<LabelCache> => {
+): Promise<Result<LabelCache, PenroseError>> => {
   const labels: LabelCache = [];
   for (const s of allShapes) {
     const { shapeType, properties } = s;
     if (shapeType === "Text" || shapeType === "TextTransform") {
       const shapeName: string = properties.name.contents as string;
       // HACK: getting type errors for not being able to resolve the Value type
-      const { body, width, height } = await tex2svg(
+      const svg = await tex2svg(
         properties.string.contents as string,
         shapeName,
         properties.fontSize.contents as string
       );
 
+      if (svg.isErr()) {
+        return err({
+          errorType: "SubstanceError",
+          tag: "Fatal",
+          message: svg.error,
+        });
+      }
+
+      const { body, width, height } = svg.value;
+
       // Instead of directly overwriting the properties, cache them temporarily
       // NOTE: in the case of empty strings, `tex2svg` returns infinity sometimes. Convert to 0 to avoid NaNs in such cases.
       const label: LabelData = {
         w: {
           tag: "FloatV",
-          contents: width === Infinity ? 0 : (width as number),
+          contents: width === Infinity ? 0 : width,
         },
         h: {
           tag: "FloatV",
-          contents: height === Infinity ? 0 : (height as number),
+          contents: height === Infinity ? 0 : height,
         },
-        rendered: body as HTMLElement,
+        rendered: body,
       };
       labels.push([shapeName, label]);
     }
   }
-  return Promise.all(labels);
```

Comment:
I'm not 100% sure if the async behavior will be the same now that `Promise.all` is gone. I guess since you are validating and pushing to the cache per step, it's kind of the same as before?

Reply:
Yep, that's my understanding; I think it's fine since it passed the typechecker?